### ISSUE TITLE
query WHERE xxx IN (:parameter) does not work with array of date

### ISSUE THREAD

**Issue by SimonHeimberg** [doctrine/doctrine2#6934]
This statement fails:
```php
$qb = $em->createQueryBuilder('Bundle:SomeEntity', 's');
$qb->where('s.date in (:dates)')->setParameter('dates', array(new DateTime(), new DateTime('NOW -2 hours')));
$qb->getQuery()->getResults();
```

with message
`Catchable Fatal Error: Object of class DateTime could not be converted to string`

but this works: `$qb->where('s.date in (:date1, :date2)')`

I like the `in (:array)` functionality. (Works with string, int, entities, ...) and miss suport of DateTime type.

---

doctrine version: v2.5.14

**Comment by kejwmen** [comment]
As you can see in dbal PR, this is related to every custom type and support for that won't be implemented in 2.x.

You can always use `array_map` as a workaround to format instances as strings like this:

```
array_map(function(\DateTimeInterface $dateTime) {
    return $dateTime->format('Y-m-d H:i:s');
});
```

**Comment by naitsirch** [comment]
@Ocramius If this won't be changed, this issue can be closed.
Maybe this should be documented somewhere, if it isn't, yet.

**Comment by fbastien** [comment]
I agree that a `in(:array)` functionality working with dates would be very handy.

Here is another workaround which is independant from date format:

```php
$criteria = [];
for($i = 0; $i < count($dates); $i++) {
    $criteria[] = $qb->expr()->eq('s.date', ":date_$i");
    $qb->setParameter("date_$i", $dates[$i]);
}
$qb->andWhere($qb->expr()->orX()->addMultiple($criteria));
```

(I'm not sure whether `Y-m-d H:i:s` is a SQL standard being recognized by any server implementation and locale settings. This may have the advantage to be more portable.)
